File: zig-updater.clj
#!/usr/bin/env bb

(ns zig-updater
  (:require [babashka.http-client :as http]
            [cheshire.core :as json]
            [clojure.string :as str]
            [babashka.fs :as fs]))

(defn split-version [version]
  (let [first-part (first (str/split version #"-"))
        second-part (second (str/split version #"-"))]
    [first-part second-part]))

(defn get-master-version [url]
    (let [zig-json (http/get url)
          data (json/parse-string (:body zig-json) true)
          master-version (get-in data [:master :version])]
    :out
    master-version))

(println (split-version (get-master-version "https://ziglang.org/download/index.json")))

(defn update-zig-spec [new-version]
  (let [spec-file-path "zig.spec"
        prev-file-path ".prev.txt"
        prev-version (split-version (slurp prev-file-path))]
    (println "Previous version:" prev-version)
    (when (not= (second prev-version) (second new-version))
      (let [content (slurp spec-file-path)
            updated-content (str/replace content (str/trim-newline (second prev-version)) (str/trim-newline (second new-version))
            )]
        (spit spec-file-path updated-content)
        (spit prev-file-path (str/join "-" new-version))))))

(let [new-version (split-version (get-master-version "https://ziglang.org/download/index.json"))]
  (update-zig-spec new-version)
  (println "Updated zig.spec with version:" new-version))

File: zig-updater.py
#!/usr/bin/env python3

import json
import os
import urllib.request

def split_version(version):
    return version.split('-')

def get_master_version(url):
    with urllib.request.urlopen(url) as response:
        data = json.loads(response.read().decode())
    return data['master']['version']

def read_prev_version():
    prev_file_path = ".prev.txt"
    with open(prev_file_path, 'r') as file:
        prev_version = file.read().strip()
    return split_version(prev_version)

def update_zig_spec(new_version):
    spec_file_path = "zig.spec"
    prev_version = read_prev_version()
    print(f"Previous version: {prev_version}")
    print(f"New version: {new_version}")

    prev_first = prev_version[0] if len(prev_version) > 0 else ''
    prev_second = prev_version[1] if len(prev_version) > 1 else '1'
    new_first = new_version[0] if len(new_version) > 0 else ''
    new_second = new_version[1] if len(new_version) > 1 else '1'

    if prev_first != new_first or prev_second != new_second:
        with open(spec_file_path, 'r') as file:
            content = file.read()
        updated_content = content.replace(f"{prev_first}", f"{new_first}")
        if prev_second is not None and new_second is not None:
            updated_content = updated_content.replace(f"%define prerelease {prev_second}", f"%define prerelease {new_second}")
            with open(spec_file_path, 'w') as file:
                file.write(updated_content)

def write_prev_version(new_version):
    prev_file_path = ".prev.txt"
    with open(prev_file_path, 'w') as file:
        file.write('-'.join(new_version) + '\n')

if __name__ == "__main__":
    url = "https://ziglang.org/download/index.json"
    new_version = split_version(get_master_version(url))
    update_zig_spec(new_version)
    print(f"Updated zig.spec with version: {new_version}")
    write_prev_version(new_version)
